redesign and make 3.1 ready (#7)
ibm-cloud-architecture/terraform-icp-ibmcloud: templates/icp-ce-with-loadbalancers/README.md

# Terraform ICP IBM Cloud

This Terraform example configurations uses the [IBM Cloud  provider](https://ibm-cloud.github.io/tf-ibm-docs/index.html) to provision virtual machines on IBM Cloud Infrastructure (SoftLayer)
and [Terraform Module ICP Deploy](https://github.com/ibm-cloud-architecture/terraform-module-icp-deploy) to prepare VSIs and deploy [IBM Cloud Private](https://www.ibm.com/cloud-computing/products/ibm-cloud-private/) version 3.1.0 or later.  This Terraform template automates best practices learned from installing ICP on IBM Cloud Infrastructure.

## Deployment overview
This template creates an environment where

>  - Cluster is deployed on private network and is accessed loadbalancers.

Reviewer: ```suggestion
 - Cluster is deployed on private network and is accessed through load balancers.
```

 - SSH access is enabled on public network
 - Most ICP services disabled (some can be activated via `terraform.tfvars` settings as described below)
 - Minimal VM sizes
 - No separate boot node

>  - No management node (can be enabled via `terraform.tfvars` settings as described below)

Reviewer: No VA node either

* The template is tested on VSIs based on Ubuntu 16.04.  RHEL is not supported in this automation.

### Using the Terraform templates

1. git clone the repository

1. Navigate to the template directory `templates/icp-ce-minimal-with-loadbalancers`

1. Create a `terraform.tfvars` file to reflect your environment.  Please see [variables.tf](variables.tf) and below tables for variable names and descriptions.  Here is an example `terraform.tfvars` file:

```
sl_username = "<my username>"
sl_api_key  = "<my api key>"
datacenter  = "dal13"
key_name    = ["my-ssh-key"]
```

1. Run `terraform init` to download depenencies (modules and plugins)

1. Run `terraform plan` to investigate deployment plan

1. Run `terraform apply` to start deployment.

### Automation Notes

#### What does the automation do
1. Create the virtual machines as defined in `variables.tf` and `terraform.tfvars`

>   - Use cloud-init to add a user `icpdeploy` with a randomly generated ssh-key

Reviewer: ```suggestion
   - Use cloud-init to add a user `icpdeploy` with a randomly generated ssh-key
```

>   - Configure a separate hard disk to be used by docker

Reviewer: ```suggestion
   - Configure a separate hard disk to be used by docker
```

2. Create security groups and rules for cluster communication as declared in [security_group.tf](security_group.tf)

> 3. Create Loadbalancers for Proxy and Control plane

Reviewer: ```suggestion
3. Create load balancers for Proxy and Control plane
```

4. Handover to the [icp-deploy](https://github.com/ibm-cloud-architecture/terraform-module-icp-deploy) terraform module as declared in the [icp-deploy.tf](icp-deploy.tf) file

#### What does the icp deploy module do
1. It uses the provided ssh key which has been generated for the `icpdeploy` user to ssh from the terraform controller to all cluster nodes to install ICP prerequisites
2. It generates a new ssh keypair for ICP Boot(master) node to ICP cluster communication and distributes the public key to the cluster nodes. This key is used by the ICP Ansible installer.

> 3. It populates the necessary /etc/hosts file on the boot node

Reviewer: ```suggestion
3. It populates the necessary `/etc/hosts` file on the boot node
```

4. It generates the ICP cluster hosts file based on information provided in [icp-deploy.tf](icp-deploy.tf)

> 5. 4. It generates the ICP cluster `config.yaml` file based on information provided in [icp-deploy.tf](icp-deploy.tf)

Reviewer: ```suggestion
5. It generates the ICP cluster `config.yaml` file based on information provided in [icp-deploy.tf](icp-deploy.tf)
```

| `private_vlan_router_hostname` | no | Private VLAN router to place all VSIs behind.  e.g. bcr01a. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. This option should be used when setting `private_network_only` to true along with `private_vlan_number` using a private VLAN that is routed with a Gateway Appliance. |
| `private_vlan_number` | no | Private VLAN number to place all VSIs on.  e.g. 1211. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. This option should be used when setting `private_network_only` to true along with `private_vlan_router_hostname`, using a private VLAN that is routed with a Gateway Appliance.|
| `public_vlan_router_hostname` | no | Public VLAN router to place all VSIs behind.  e.g. fcr01a. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. |
| `public_vlan_number` | no | Public VLAN number to place all VSIs on.  e.g. 1211. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. |
| `icppassword` | no | ICP administrator password.  One will be generated if not set. |

> 

Reviewer: ```suggestion
| `deployment` | no | Identifier prefix added to the host names of all your infrastructure resources for organising/naming ease |
```
